Add script to look for all methods

JanJecmen opened the pull request:
Putting this here for discussion.

This should go over a given project and all its dependencies, and find all methods the project defines.
Note, it also looks for direct definitions in external modules, eg. `Base.push!() = 1`.

Basically, it enumerates all top-level modules of packages reachable through the dependency graph (including `Base` as a special case), and looks for all functions in each module (also recursively in submodules).
Then it gets each function's method table, and filters methods coming from the project (implemented as a subdirectory check as of now)

ulysses4ever commented on scripts/find_all_methods.jl:
```diff
@@ -0,0 +1,148 @@
+# Report
+
+open(joinpath(out_dir, "method_report.md"), "w") do f
+    write(f, "# Report of methods found for project at `$project_dir`\n")
```
Should be `print`.

ulysses4ever commented on scripts/find_all_methods.jl:
```diff
@@ -0,0 +1,148 @@
+    OUT_DIR ....... directory where output is written, also a fresh depot is maintained here
+
+Notes:
+    - We look at the project's main module, all its transitive dependencies' toplevel modules, and the Base module
```
Apart from `Base`, you need to special-case `Core` (not sure how often people extend stuff from `Core` but they probably do it sometimes). This is documented https://docs.julialang.org/en/v1/manual/modules/#Default-top-level-definitions-and-bare-modules

JanJecmen replied:
I thought you can't extend anything from Core as it contains builtin functions?

JanJecmen replied:
I see, not only..

ulysses4ever replied:
Right. You need to distinguish functions (names) from methods (implementations). Core can define a function name foo and say that there's a built-in implementation of foo for the input of type Bar. Later on, a user module can add another, normal (non built-in) method for the same function (name) but different input type.

ulysses4ever commented:
Good stuff. I have couple comments, but generally this is probably the way to go.

I'm always worried when you have to work via filesystem: there are many kinds of issues that you can run into. It'd be good, at least, to be very explicit why you need to access the filesystem. It seems, currently, it's used to "Filter only methods from our project". I think, this could be factored into two parts: (1) filter only methods from the given module / list of modules, and (2) list all modules of the project. (1) doesn't need filesystem. (2) we already discussed separately and may have a solution for? It could access the filesystem, if there's no way around it, but at least it doesn't do much besides, and so it should be easier to maintain than a monolithic code solving (1) and (2) together.

When searching dependencies, the algorithm assumes that every package  defines only one top-level module and that module has the same name as the package. [Based on documentation](https://docs.julialang.org/en/v1/manual/code-loading/), it's looks like a safe assumption. I tried to `Pkg.generate` a fresh package `A` and add a top-level module `B` there. `B` doesn't seem to be available after adding the package to the current environment. I thought it would but this way it makes our life easier.

General comment on code organization: it's better to not add too much stuff in `scripts` because we don't run tests there at all. And code without tests is very hard to maintain. Maybe putting it in `src` and creating a top-level function that "does the job" is possible here? I'm not sure. I admit the distinction between `scripts` and `src` is subtle. There were several reasons for this distinction. One is to have little dependency footprint for the core functionality. Another is to have a home for bash scripts. Maybe, for the former, at least, I should have created a separate Julia package.

Finally, there's some duplication here, e.g. the loop over `names` in `process_module_aux` looks very similar to what we have in `is_stable_module_aux`. Also, `get_module` may have been implemented somewhere. Maybe, this all could be factored out...

JanJecmen committed:
Reworked function discovery

This helps detect external method definitions. We now look for functions in the given module, as well as all of user-provided extra modules, and then filter all methods of all the discovered functions by whether they come from the module of interest.

A good option is to pass `Base.loaded_modules_array()` as the extra modules argument, as this will return a vector of all the modules loaded currently in the julia session. This includes Base, Core, the module of interest, and all of its transitive dependencies. Any extra modules that also happen to be loaded are processed too, but this doesn't seem to affect performace much.

ulysses4ever commented on test/runtests.jl:
```diff
@@ -250,8 +250,32 @@ end
     chks = is_stable_module(ImportBase)
     @test length(chks) == 1
     @test chks[1].check == Stb(1)
+    chks = is_stable_module(NonImportBase; extra_modules=[Base])
```
It looks like `[Base]` could be a better default than `[]`?

JanJecmen replied:
I thought it's better to leave this explicit, i.e., by default you only get functions you find in `m`. But I can change it if you want - maybe `[Base, Core]` would make sense then?

ulysses4ever replied:
I love the description in README, thanks! Why not `Base.loaded_modules_array()` as the default? It seems "more correct" and not too expensive?..

ulysses4ever reviewed (approved):
Sweet!

In general, it'd be nice to add something to README (and probably extend the source comment) about the `extra_modules` parameter, which currently looks rather mysterious. What should I pass if I want "really all" methods? Direct dependencies? A transitive closure of dependencies? This could benefit from some discussion with examples, I think.

ulysses4ever commented:
Also, I seem to have missed when this PR switched from Draft to Ready. Maybe leave a comment next time when switching.

ulysses4ever commented:
Oh, I see now that you added some explanation in the commit message. We definitely want it in the source too! (comments at least, maybe readme)

JanJecmen committed:
Document the new `extra_modules` feature

JanJecmen commented:
Added explanations to both a comment and readme. Let me know if you think this is sufficient.
